spec/index.bs

```diff
@@ -3693,21 +3693,111 @@ GPURenderPipeline includes GPUPipelineBase;
+                - initialize the |vertexList| with |drawCall|.indexCount integers
+                - assign the |vertexList| item |i| to be [$fetch index$](|i| + |drawCall|.firstIndex, |indexBuffer|.buffer, |indexBuffer|.offset, |indexBuffer|.format) + |drawCall|.baseVertex
+            If |vertexState|.{{GPUVertexStateDescriptor/indexFormat}} is `null`:
+                - initialize the |vertexList| with |drawCall|.vertexCount integers
+                - assign the |vertexList| item |i| to be |drawCall|.firstVertex + |i|
+
+        1. Process vertices. Each vertex |vertexIndex| in the |vertexList|,
+            in each instance of index |rawInstanceIndex| in range from 0 to |drawCall|.instanceCount - 1, inclusive,
+            is processed separately:
+            1. Let |instanceIndex| be |rawInstanceIndex| + |drawCall|.baseInstance.
+            1. For each non-`null` |vertexBufferDesc| under index |i|
+                in the list of |vertexState|.{{GPUVertexStateDescriptor/vertexBuffers}}:
+                1. Let |vertexBuffer| and |vertexBufferOffset| be the buffer and offset in |vertexBuffers| bindings at slot |i|
+                1. Let |vertexElementIndex| be dependent on |vertexBufferDesc|.{{GPUVertexBufferLayoutDescriptor/stepMode}}:
+                    <dl class="switch">
+                        : {{GPUInputStepMode/"vertex"}}
+                        :: |vertexIndex|
+                        : {{GPUInputStepMode/"instance"}}
+                        :: |instanceIndex|
+                    </dl>
+                1. For each |attributeDesc| in |vertexBufferDesc|.{{GPUVertexBufferLayoutDescriptor/attributes}}:
+                    1. Let |attributeOffset| be |vertexBufferOffset| +
+                        |vertexElementIndex| * |vertexBufferDesc|.{{GPUVertexBufferLayoutDescriptor/arrayStride}} +
+                        |attributeDesc|.{{GPUVertexAttributeDescriptor/offset}}.
+                    1. Load the attribute |data| of format |attributeDesc|.{{GPUVertexAttributeDescriptor/format}}
+                        from |vertexBuffer| starting at offset |attributeOffset|.
+                    1. Convert the |data| into a shader-visible format.
+                        <div class="example">
+                        An attribute of type {{GPUVertexFormat/"uchar2norm"}} will be converted from
+                            2 bytes of fixed-point unsigned 8-bit integers into 2 floating-point values
+                            as `vec2<f32>` in WGSL.
+                        </div>
+                    1. Bind the |data| to vertex shader input
+                        location |attributeDesc|.{{GPUVertexAttributeDescriptor/shaderLocation}}.
+            1. For each {{GPUBindGroup}} group at |index| in |bindGroups|:
+                1. Bind the group of resources visible to vertex shader group |index|.
+            1. Invoke vertex shader in |descriptor|.{{GPURenderPipelineDescriptor/vertexStage}}.
```

Comment:
We might even model vertex shaders as functions of bindings, inputs(builtins and attributes) that returns outputs (varyings and builtins) while potentially doing side effects. Then call that explicitly.

The other thing is that we should be explicit as to which steps happen in parallel. Maybe we could say that the function executes asynchonously. And in general we should explain what parts of rendering happen on one another, and which parts are done in parallel. (simlarly we could describe that with the concept of promises maybe?)

Reply:
At some point, we need to say that we give the driver control to invoke our shader. Whether here, or inside another algorithm. Is there a benefit for wrapping this "invoke xxx" inside a separate algorithm?

Reply:
I'm not sure what's best. What I feel stronger about is that the synchronization guarantees and the parallel nature of GPUs should be visible in the spec. (like VS and FS for a render pass running in parallel or out of order for different triangles)

Reply:
Yes, I totally agree. The PR now has a bit of wording to be clear about that. We'll add more.